torch/csrc/api/include/torch/detail/any.h

```diff
@@ -0,0 +1,138 @@
+#pragma once
+
+#include <torch/csrc/utils/memory.h>
+#include <torch/csrc/utils/variadic.h>
+
+#include <ATen/Error.h>
+
+#include <memory>
+#include <typeinfo>
+#include <utility>
+
+namespace torch {
+class Any;
+namespace detail {
+template <typename T>
+Any make_any(T&& value);
+} // namespace detail
+} // namespace torch
+
+namespace torch {
+/// A simplified implementation of `std::any` or `boost::any` which stores a
+/// type erased object, whose concrete value can be retrieved at runtime by
+/// checking if the `typeid()` of a requested type matches the `typeid()` of the
+/// object stored. It is simplified in that it does not handle copying, as we do
+/// not require it for our use cases. Moves are sufficient.
+class Any {
+ public:
+  /// Default construction is disallowed. Thus our invariant: an `Any` contains
+  /// an object at all times, from its construction, to its destruction.
+  Any() = delete;
+
+  /// Move construction and assignment is allowed, and follows the default
+  /// behavior of move for `std::unique_ptr`.
+  Any(Any&&) = default;
+  Any& operator=(Any&&) = default;
+
+  /// Copy is disallowed, because we don't need it.
+  Any(const Any& other) = delete;
+  Any& operator=(const Any& other) = delete;
+
+  /// Returns the value contained in the `Any` if the type passed as template
+  /// parameter matches the type of the object stored, and returns a null
+  /// pointer otherwise.
+  template <typename T>
+  T* try_get() {
+    static_assert(
+        !std::is_reference<T>::value,
+        "Any stores decayed types, you cannot cast it to a reference type");
+    static_assert(
+        !std::is_array<T>::value,
+        "Any stores decayed types, you must cast it to T* instead of T[]");
+    if (typeid(T).hash_code() == type_info().hash_code()) {
```

Comment:
I don't think you can rely on `hash_code` being unique.

Reply:
So cppreference says:
> There is no guarantee that the same std::type_info instance will be referred to by all evaluations of the typeid expression on the same type, although std::type_info::hash_code of those type_info objects would be identical, as would be their std::type_index.

At the same time, libc++ just compares the values returned by `typeid()`: https://github.com/llvm-mirror/libcxx/blob/master/include/any#L166

`hash_code()` sounds safer to me?

Reply:
Possibly clang gives this guarantee but the standard doesn't. We shouldn't rely on that either then.
`hash_code` is wrong the other way round. While comparing `typeid` directly might give you false negatives, `hash_code` might give you false positives. Can you use the `std::type_index` for comparison or does that need RTTI? If yes, how does boost::any solve it?

Reply:
`std::type_index` has the same guarantees as `hash_code`. The above quote is from cppreference, so it should be standard?

Reply:
Quote from cppreference:

> No other guarantees are given: type_info objects referring to different types may have the same hash_code (although the standard recommends that implementations avoid this as much as possible), and hash_code for the same type can change between invocations of the same program.